libraries/AP_Arming/AP_Arming.cpp

```diff
@@ -850,6 +850,116 @@ bool AP_Arming::camera_checks(bool display_failure)
     return true;
 }
 
+// request an auxiliary authorisation id.  This id should be used in subsequent calls to set_aux_auth_passed/failed
+// returns true on success
+bool AP_Arming::get_aux_auth_id(uint8_t& auth_id)
+{
+    WITH_SEMAPHORE(aux_auth_sem);
+
+    // check we have enough room to allocate another id
+    if (aux_auth_count >= aux_auth_count_max) {
+        aux_auth_error = true;
+        return false;
+    }
+
+    // allocate buffer for failure message
+    if (aux_auth_fail_msg == nullptr) {
+        aux_auth_fail_msg = (char *)calloc(aux_auth_str_len, sizeof(char));
+        if (aux_auth_fail_msg == nullptr) {
+            aux_auth_error = true;
+            return false;
+        }
+    }
+    auth_id = aux_auth_count;
+    aux_auth_state[auth_id] = AuxAuthStates::AUX_AUTH_STATE_NO_RESPONSE;
+    aux_auth_count++;
+    return true;
+}
+
+// set auxiliary authorisation passed
+void AP_Arming::set_aux_auth_passed(uint8_t auth_id)
+{
+    WITH_SEMAPHORE(aux_auth_sem);
+
+    // sanity check auth_id
+    if ((auth_id >= aux_auth_count_max) || (aux_auth_state[auth_id] == AuxAuthStates::AUX_AUTH_STATE_UNASSIGNED)) {
+        return;
+    }
+
+    aux_auth_state[auth_id] = AuxAuthStates::AUX_AUTH_STATE_PASSED;
+}
+
+// set auxiliary authorisation failed and provide failure message
+void AP_Arming::set_aux_auth_failed(uint8_t auth_id, const char* fail_msg)
+{
+    WITH_SEMAPHORE(aux_auth_sem);
+
+    // sanity check auth_id
+    if ((auth_id >= aux_auth_count_max) || (aux_auth_state[auth_id] == AuxAuthStates::AUX_AUTH_STATE_UNASSIGNED)) {
+        return;
+    }
+
+    // update state
+    aux_auth_state[auth_id] = AuxAuthStates::AUX_AUTH_STATE_FAILED;
+
+    // store failure message if this authoriser has the lowest auth_id
```

Comment:
Why bother with storing the lowest ID first? The order in which scripts get loaded here is not well defined, so there is nothing special about ID 0 before ID 1. Why not just always store the most recent failure one here? That has the merits of being both less code and stores the failure that we know the most about. If you had a script that was checking GPS information at 0.2Hz and a different script monitoring the status of some RC channel at 1Hz as an example, the GPS one may be true now and just waiting for the script to run, but if the RC channel is asserting its failed the most recently then this is the most likely thing for a user to need to fix.

Reply:
I think I'll leave this one as it is which leads to the first authoriser's message being displayed until it passes.  If we don't do this then we end up with messages from various authorisers being somewhat randomly reported.  I.e. it could easily display the failure from the first authoriser, then a moment later display the 2nd authoriser's failure message, then back to the first.

I think we should either treat all the authorisers as if they're part of the same subsystem (i.e. display the first message until it passes, then the 2nd, etc) or we treat them as if they're all different subsystems in which case we go back to the original plan of having separate message buffers for each authoriser.